### PR TITLE
dflash: AngelSpec DFly drafter support

### PR DESCRIPTION
**Pull request by bri-prism**
Replaces #146, re-branched off `prism`. That PR targeted `prism-v7`, which is frozen; per the guidance on #169 the retarget is done by re-branching rather than by changing the base in place, since an in-place flip drags in every `prism-v7`-only commit under the branch.

Same seven commits, cherry-picked clean onto `prism` with no conflicts. 13 files, +816/-15, matching #146 exactly. Eleven of the thirteen files are byte-identical to #146's head `920c50d1d`; the two that differ (`gguf-py/gguf/constants.py`, `tests/CMakeLists.txt`) differ only where `prism` carries PTQ1_0 (#148) that `prism-v7` never got. Every DFly addition is intact.

CI note: the Windows job on #146 had one test failure, `unit/test_completion.py::test_completion_unified[256-4-n_predict_vals3-expected_success3]` (`assert 500 == 200`), from KV-cache exhaustion, with 117 passed. That run is from 2026-09-04 against the frozen base. This PR gets a fresh run on `prism`; if the same test fails again it needs a real look rather than being written off as a flake.

---

## What

Adds AngelSpec DFly drafter support to the `dflash` arch, so a DFly checkpoint can be
converted and used as a speculative drafter. Verified against the public reference pairing:
Qwen3-8B as target and `AngelSlim/Qwen3-8B-DFly-Block8` as drafter.

Also fixes one bug in `common/speculative.cpp` that is independent of DFly and could be
taken on its own (second commit).

## Why

DFly extends the DFlash draft path with two pieces we did not have:

1. **Per-draft-layer context fusion.** DFlash projects the captured target features once and
   injects that single context into every draft layer. DFly keeps that shared projection as a
   base context, then adds a per-draft-layer softmax-weighted mix of the raw per-target-layer
   features, so each draft layer gets its own context.
2. **A TreeFlash predecessor correction** applied before the target head, chained across the
   block: position i's logits come from the draft hidden state at row i corrected by the
   embedding of the token drafted at row i-1.

The expensive prerequisite was already here. `dflash` already reads a `target_layers` array
and sizes its encoder input as `|target_layers| * n_embd`, which is exactly the concatenated
feature layout DFly consumes, so multi-layer capture needed no work.

## How

DFly is detected from the presence of the `layer_fusion` tensor rather than from a metadata
key, so a DFly export cannot quietly load as plain DFlash. A checkpoint that declares both a
DFly fusion and a DSpark Markov head is rejected with that reason instead of failing later on
a missing tensor. Reduced draft vocabularies are rejected, since the reference chain runs the
full target head.

Nothing was needed in the draft reader. DFly reports no Markov head, so it already takes the
DFlash path (block rows 1 through n-1) and the plain per-token probability cut.

Two width contracts have to widen together for the fused context to survive its round trip
through the host, and this is the part worth reviewing:

* `n_embd_out` sizes the encoder's nextn output. `common/speculative.cpp` was sizing every
  buffer on that path with `llama_model_n_embd(model_dft)` instead. Those two are equal for
  any drafter that emits one context row per token, so that commit is a no-op for DFlash and
  DSpark, but they are not equal for DFly and the batch was under-allocated.
* `n_embd_inp` sizes the embd batch that feeds the context back in. A dflash draft context is
  not MTP-typed, so `llama_context::decode` takes the `n_embd_inp()` branch. Setting only
  `n_embd_out` leaves the decoder reading one layer's context and n_layer-1 layers of junk.

## Measurements

M5 Pro, Metal, bf16 target and bf16 drafter, block 8, greedy, 64 predicted tokens, through
`llama-server`, with the arms interleaved A-B-A-B:

| arm | tok/s | acceptance |
|---|---|---|
| autoregressive baseline | 18.35, 18.42, 18.50, 18.52 | n/a |
| DFly | 38.53, 38.65, 38.18, 38.00 | 52.7 percent (49/93) |
| DFly, correction disabled | 30.89 | 21.0 percent (37/176) |

That is 2.08x against a same-path baseline whose four readings spread 0.9 percent. Mean
accepted length is 4.50 of a block of 8, which puts an 8-row verify at about 2.16 plain decode
steps. That is the right order for a bandwidth-bound bf16 8B target, so the mechanism accounts
for the size of the speedup rather than just accompanying it.

The correction is load-bearing: disabling it drops acceptance from 52.7 to 21.0 percent.
`LLAMA_DFLY_NO_CHAIN=1` exists to make that A/B measurable and is not a normal setting.

## Testing

* Generated text is byte identical to the autoregressive baseline at greedy in every
  speculative arm. That is the guarantee speculative decoding owes, and it is what would catch
  a mis-indexed block or a corrupted verify.
* `tests/test-dfly-fusion.cpp` pins the fusion math and the resulting memory layout against an
  independent scalar reference. The axis handling needs two reshape and permute round trips,
  and a transposed axis would otherwise still load and still draft plausibly. The test also
  asserts the per-layer contexts actually differ, so a collapsed broadcast fails.
* `tests/gen-tiny-dfly.py` builds a tiny random DFly GGUF for loader smoke tests, including
  the rejection paths above.

## Notes for review

The conversion path covers `Qwen3DFlyModel` and the `Qwen3DSparkDFlareV2Model` alias, and
refuses a config whose `target_layer_ids` exceed its stated target depth. The published
metadata of the reference checkpoint does exactly that on its main branch, and it loads clean
and then mis-projects, so pin a known good revision when converting.

### CODE REVIEW

**Comment by bri-prism** [comment]
CI results on the new base. Seven checks resolved, all passing; four still running at the time of writing.

| check | result |
|---|---|
| ubuntu | pass, 11m32s |
| windows | pass, 15m28s |
| CodeQL | pass |
| Analyze (actions) | pass |
| Analyze (javascript-typescript) | pass |
| Analyze (python) | pass |
| labeler | pass |
| Analyze (c-cpp) | running |
| Lint | running |
| python type-check (x2) | running |

**The Windows failure from #146 did not reproduce.** That run had `unit/test_completion.py::test_completion_unified[256-4-n_predict_vals3-expected_success3]` failing on `assert 500 == 200`, behind a KV-cache exhaustion (`failed to find a memory slot for batch of size 1`, then `Context size has been exceeded`), with 117 passed. The full server suite passes here.

Stating the limit of that: this is one green run on a different base, not proof the earlier failure was a flake. Two things changed at once, the base moved from the frozen `prism-v7` to `prism` and the run happened ten days later, so this does not isolate a cause. What it does establish is that the failure is not reproducible from this branch on `prism`, which is the branch being proposed for merge.

Also worth noting for anyone comparing against #146's check list: the `Lint` and `python type-check` jobs there are recorded as failures with 24h runtimes, so they hung rather than failed on anything in the diff. They are running normally here and I will follow up if they do not resolve.

File: conversion/qwen.py [hunk 1]
```diff
@@ -729,6 +729,129 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
...
+                f"DFly target_layer_ids {layer_ids} exceed target_num_hidden_layers {target_layers}. "
+                "The config metadata does not describe the weights -- pin a known-good revision "
+                "(the reference checkpoint's is 5712926)."
+            )
+
+        if int(hp.get("target_hidden_size", hp["hidden_size"])) != int(hp["hidden_size"]):
+            raise ValueError(
+                "DFly residual fusion requires target_hidden_size == hidden_size, got "
+                f"{hp.get('target_hidden_size')} vs {hp['hidden_size']}."
+            )
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/qwen.py]: Fixed in 844b8ce74. The check now prefers the target model's own hidden_size, which is already read for the other target_* comparisons, and only falls back to the declared value and then to the draft width when there is no target config to read. An omitted target_hidden_size no longer compares a value against itself.

On the second reference: line 803 in your numbering is the hidden_correction_type / dspark_bonus_anchor defaults, and neither has this shape. Their defaults are a fixed string and a bool, not the same quantity the comparison is against, so there is nothing to mis-compare. If you meant a different site, point me at it and I will look again.

File: src/models/dflash.cpp [hunk 2]
```diff
@@ -16,6 +16,21 @@ void llama_model_dflash::load_arch_hparams(llama_model_loader & ml) {
...
 
+    // AngelSpec DFly fuses the target context once per DRAFT layer, so the encoder emits
+    // [n_embd, n_layer] per token instead of a single [n_embd] row. Detected from the fusion
+    // tensor rather than a KV so a DFly export cannot load as plain DFlash.
+    if (ml.get_tensor_meta("layer_fusion")) {
+        // Both ends of the round-trip must widen: the encoder emits n_layer contexts
+        // (n_embd_out) and the decoder's embd batch consumes them (n_embd_inp). Setting only
+        // the former is not enough -- a dflash draft context is not MTP-typed, so
+        // llama_context::decode sizes an embd batch from n_embd_inp(), and leaving that at
+        // n_embd hands the graph one layer's context and n_layer-1 layers of junk.
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply src/models/dflash.cpp]: Trimmed to the invariant in 844b8ce74: two lines, keeping only why both ends have to widen.

File: src/models/dflash.cpp [hunk 3]
```diff
@@ -389,6 +490,138 @@ static void build_dspark_markov_head(llm_graph_context & g, const llama_model &
...
+                g.build_norm(h_i,       model.dfly_hc_hidden_norm, NULL, LLM_NORM_RMS, -1),
+                g.build_norm(prev_embd, model.dfly_hc_embed_norm,  NULL, LLM_NORM_RMS, -1), 0);
+
+        ggml_tensor * delta = g.build_ffn(z,
+                model.dfly_hc_up,   NULL, NULL,
+                model.dfly_hc_gate, NULL, NULL,
+                model.dfly_hc_down, NULL, NULL,
+                NULL, LLM_FFN_SILU, LLM_FFN_PAR, -1);
+
+        ggml_tensor * col = g.build_lora_mm(output, ggml_add(ctx0, h_i, delta), output_s);
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply src/models/dflash.cpp:603]: The waste is real and I want to fix it, but not in this PR, and here is why.

base is res->t_logits, which the generic head already built and expanded into the graph before this hook runs, so nothing inside the hook can remove those rows. Skipping them means not building the full head in the first place, which has to happen in the decoder before the hook, while several of the hook's bail-out paths (non-block batch, oversized block, anchor out of range) still need full logits for every row. That is a decoder change with a correctness edge, not a local rewrite.

The placeholder variant is also not free here: the anchor column is what keeps one output row per ubatch token, and common/speculative.cpp reads rows 1..n-1, so I would want a measurement before relying on the consumer skipping row 0 rather than on it being correct.

Opening it as a follow-up with the profile attached, so the change is justified by a number rather than by a row count. If you would rather see it in this PR, say so and I will do it here instead.

**Comment by bri-prism** [comment]
Follow-up for the double head projection in the DFly correction head is #176, with the decoder-side reason it cannot be fixed inside the hook and the profile I want before restructuring.
